### PR TITLE
Add drop_remainder & redistribute to torch.chunk and drop_remainder for torch.split

### LINKED ISSUE #60531: Enhanced torch.chunk and torch.split
## 🚀 Feature

Add `drop_remainder: bool = True` support to `torch.chunk` and `torch.split`
+ similar in function to `drop_remainder` in `torch.utils.data.DataLoader`
+ If the length of the dimension is not perfectly divisible, drop the remaining elements from the returned arrays.

Add `redistribute: bool = True` support to `torch.chunk` and `torch.split`
+ Spread the elements evenly across the chunks/splits if the length of the dimension is not perfectly divisible by the number of chunks or split size.
+ The length of returned tensors differs by at most one element.

## Motivation

Often it is desirable to have evenly sizes arrays, or approximately even sized arrays.

A combination of `drop_remainder` and `redistribute` would enable new intuitive ways to use `torch.split` and `torch.chunk`.

## Pitch

**redistribute**

Add the parameter `redistribute` to `torch.chunk` and `torch.split` which spreads the remainder evenly across the returned tensors. For example:

```python3
# CURRENT:
torch.arange(7).chunk(3, dim=-1, redistribute=False)
>>> (tensor([0, 1, 2]), tensor([3, 4, 5]), tensor([6]))

# NEW:
torch.arange(7).chunk(3, dim=-1, redistribute=True)
>>> (tensor([0, 1, 2]), tensor([3, 4]), tensor([5, 6]))
```

```python3
# CURRENT:
torch.arange(7).split(3, dim=-1, redistribute=False)
>>> (tensor([0, 1, 2]), tensor([3, 4, 5]), tensor([6]))

# NEW:
torch.arange(7).split(3, dim=-1, redistribute=True)
>>> (tensor([0, 1, 2]), tensor([3, 4]), tensor([5, 6]))
```

**drop_remainder**

Add the parameter `drop_remainder` to `torch.chunk` and `torch.split` which excludes the remainder elements from being returned, and so all the returned tensors will be the same size. For example:

```python3
# CURRENT:
torch.arange(7).chunk(3, dim=-1, drop_remainder=False)
>>> (tensor([0, 1, 2]), tensor([3, 4, 5]), tensor([6]))

# NEW:
torch.arange(7).chunk(3, dim=-1, drop_remainder=True)
>>> (tensor([0, 1]), tensor([2, 3]), tensor([4, 5]))
```

```python3
# CURRENT:
torch.arange(7).split(3, dim=-1, drop_remainder=False)
>>> (tensor([0, 1, 2]), tensor([3, 4, 5]), tensor([6]))

# NEW:
torch.arange(7).split(3, dim=-1, drop_remainder=True)
>>> (tensor([0, 1, 2]), tensor([3, 4, 5]))
```

**drop_remainder & redistribute interaction**

`drop_remainder` should be applied before `redistribute`:
- thus if both `drop_remainder=True`, then there is never a need to redistribute elements so in this case `redistribute` can be ignored.

## Alternatives

New functions that provide similar behavior or act over the outputs from `chunk` or `split`, however this might be confusing in the first case or inefficient in the second case.

Possible synonyms for `drop_remainder` include:
+ `drop_last` matching `torch.utils.data.DataLoader`, but this does not convey the correct behaviour in all cases.

Possible synonyms for `redistribute` include
+ `redistribute_remainder`
+ `distribute_remainder` 
+ `spread_remainder`
+ `distribute`
+ `spread`

The `_remainder` does not describe the movement of the remainder/last elements well, as internal elements are shifted too.

## Additional context

N/A

### PR DESCRIPTION
**Pull request by andreigh**
Implements the exact expected behavior described [here](https://github.com/pytorch/pytorch/issues/60531#issuecomment-870986104).

Fixes #60531

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @penguinwu @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan

### CODE REVIEW

**Comment by nmichlo** [comment]
Awesome!

**Comment by andreigh** [comment]
> Awesome!

@nmichlo there's a lot of tests failing now I see with:

`Couldn't find an operator for aten::chunk(Tensor self, int chunks, int dim) -> Tensor[]. Do you have to update a set of hardcoded JIT ops?
`
and similar for aten::split. Would this PR break a lot of stuff, because it changes the signature of an operator ? Is there a way to make it backward compatible (because the new parameter has a default value) ? I can go forward and fix all the tests, but wondering if this will break users' serialized models or something.

Thank you for any help on this

**Comment by andreigh** [comment]
@pytorchbot label "suppress-bc-linter"

Warning: Function split: drop_remainder was added and is now required

**Comment by andreigh** [comment]
> I was pointed from another pr [#109966 (comment)](https://github.com/pytorch/pytorch/pull/109966#issuecomment-1737666520) and was told that xla test is failing. Is that still the case?

Thank you for assisting with this. This PR changes the signature of tensor.split, and the PyTorch XLA code uses the already existing signature. Thus, the CI compile step (linux-focal-py3_8-clang9-xla / test) for this PR [fails](https://github.com/pytorch/pytorch/actions/runs/6229091072/job/16907283210?pr=107484) because it looks for the old signature.

**Comment by zpcore** [comment]
Thanks for pointing out the issue. We will take a look today.

**Comment by qihqi** [comment]
> > I was pointed from another pr [#109966 (comment)](https://github.com/pytorch/pytorch/pull/109966#issuecomment-1737666520) and was told that xla test is failing. Is that still the case?
> 
> Thank you for assisting with this. This PR changes the signature of tensor.split, and the PyTorch XLA code uses the already existing signature. Thus, the CI compile step (linux-focal-py3_8-clang9-xla / test) for this PR [fails](https://github.com/pytorch/pytorch/actions/runs/6229091072/job/16907283210?pr=107484) because it looks for the old signature:
> 
> ```
> torch_xla/csrc/aten_xla_type.cpp:2873:25: error: no declaration matches 'std::vector<at::Tensor> torch_xla::XLANativeFunctions::split_copy(const at::Tensor&, int64_t, int64_t)'
> 
> no declaration matches 'std::vector<at::Tensor> torch_xla::XLANativeFunctions::split_with_sizes_copy(const at::Tensor&, at::IntArrayRef, int64_t)'
> ```

noob Q: Any reason for not adding a default val of the newly added argument? That seems to be non-BC-breaking.

**Comment by andreigh** [comment]
> > > I was pointed from another pr [#109966 (comment)](https://github.com/pytorch/pytorch/pull/109966#issuecomment-1737666520) and was told that xla test is failing. Is that still the case?
> > 
> > 
> > Thank you for assisting with this. This PR changes the signature of tensor.split, and the PyTorch XLA code uses the already existing signature. Thus, the CI compile step (linux-focal-py3_8-clang9-xla / test) for this PR [fails](https://github.com/pytorch/pytorch/actions/runs/6229091072/job/16907283210?pr=107484) because it looks for the old signature:
> > ```
> > torch_xla/csrc/aten_xla_type.cpp:2873:25: error: no declaration matches 'std::vector<at::Tensor> torch_xla::XLANativeFunctions::split_copy(const at::Tensor&, int64_t, int64_t)'
> > 
> > no declaration matches 'std::vector<at::Tensor> torch_xla::XLANativeFunctions::split_with_sizes_copy(const at::Tensor&, at::IntArrayRef, int64_t)'
> > ```
> 
> noob Q: Any reason for not adding a default val of the newly added argument? That seems to be non-BC-breaking.

I'm setting [the default value](https://github.com/pytorch/pytorch/pull/107484/files#diff-2f3dbd85efb9b5172f2264eedd3be47dd765e6ab7cc8bf3ade5e62c28ae35991), it still seems that it wants the native op to have the exact parameters.

**Comment by zpcore** [comment]
Hi @andreigh , we have created the XLA PR [#5655](https://github.com/pytorch/xla/pull/5655) (hash: 22b777961074110fb8101c16a2eda7ddddbf1858). Please update your PR with the fix commit hash once the XLA test CI is complete. Thanks!

**Comment by andreigh** [comment]
> Hi @andreigh , we have created the XLA PR [#5655](https://github.com/pytorch/xla/pull/5655) (hash: 22b777961074110fb8101c16a2eda7ddddbf1858). Please update your PR with the fix commit hash once the XLA test CI is complete. Thanks!

Erm, how can I do this ? "update your PR with the fix commit hash"

**Comment by lsy323** [comment]
Hi @andreigh, our XLA change is ready to land. Can you please update the xla pin in your PR, then we can merge once CI is green.

here is an example how to update xla pin https://github.com/pytorch/pytorch/pull/87817/files#diff-4cfabe3d68e5f2ee83dc59768c3a31246cdf8e0672229a956dad08101e3919f9

Thank you

**Comment by justinchuby** [comment]
Could you enable/update OpInfo tests in https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_methods_invocations.py as well? Thanks!

File: test/onnx/test_fx_op_consistency.py [hunk 1]
```diff
@@ -275,6 +275,10 @@
             "Chunk", "uint8, int8, int16, float16"
         ),
     ),
+    xfail(
```
Comment (justinchuby) [inline test/onnx/test_fx_op_consistency.py:278]:
```suggestion
    xfail_torchlib_forward_compatibility(
```

Could you rebase on main and use `xfail_torchlib_forward_compatibility` instead? Thanks!

File: test/onnx/test_fx_op_consistency.py [hunk 2]
```diff
@@ -275,6 +275,10 @@
...
+        "chunk", dtypes=onnx_test_common.BOOL_TYPES + onnx_test_common.INT_TYPES + onnx_test_common.FLOAT_TYPES,
```
Comment (justinchuby) [inline test/onnx/test_fx_op_consistency.py:279]:
```suggestion
        "chunk",
```

fine to not include dtypes if it's everything

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 3]
```diff
@@ -2110,9 +2110,14 @@ class ShapePropagator : public PropertyPropBase {
...
-                   "aten::chunk(Tensor self, int chunks, int dim) -> Tensor[]",
-                   /*const_inputs=*/{attr::chunks, attr::dim})) {
+    } else if (
+        node->matches(
+            "aten::chunk(Tensor self, int chunks, int dim, bool redistribute, bool drop_remainder) -> Tensor[]",
+            /*const_inputs=*/
+            {attr::chunks,
+             attr::dim,
+             attr::redistribute,
+             attr::drop_remainder})) {
```
Comment (andreigh) [inline torch/csrc/jit/passes/shape_analysis.cpp:2120]: The algorithm here seems incorrect to me: the sizes for all chunks (except last) is set to 

`input_type->sizes()[dim] / chunks`, 

and the size of the last chunk is set to 

`input_type->sizes()[dim] % chunks`.

This is incorrect, if sizes[dim] is 5 and chunks 2, the sizes of the outputs will be 2 and 1.
Can anyone give me some suggestions what to do here ? Thank you

Comment (andreigh) [reply torch/csrc/jit/passes/shape_analysis.cpp:2120]: I can totally change the logic, my comment was to point out the current existing logic doesn't seem correct to me (and it was different from the current logic in torch.chunk). I don't really understand how the difference in the behaviours went unnoticed.

Is there a way to adjust the number of node->outputs() ? The correct behavior requires computing the number of outputs based on redistribute & drop_remainder (see https://github.com/pytorch/pytorch/issues/60531#issuecomment-870986104).

Thank you !

Comment (zpcore) [reply torch/csrc/jit/passes/shape_analysis.cpp:2120]: Sorry, I thought you added the drop_remainder feature. Authors from Pytorch may know the answer. We will update coordinately once we have the solution. Thanks!

Comment (andreigh) [reply torch/csrc/jit/passes/shape_analysis.cpp:2120]: I added the drop_remainder feature, but the existing code, before my changes, doesn't look right to me :)

File: aten/src/ATen/native/TensorShape.cpp [hunk 4]
```diff
@@ -2614,7 +2637,7 @@ std::vector<Tensor> dsplit(const Tensor& self, int64_t split_size) {
   return at::tensor_split(self, split_size, 2);
 }
 
-std::vector<Tensor> split_with_sizes(const Tensor& self, IntArrayRef split_sizes, int64_t dim) {
+std::vector<Tensor> split_with_sizes(const Tensor& self, IntArrayRef split_sizes, int64_t dim, bool drop_remainder) {
```
Comment (zpcore) [inline aten/src/ATen/native/TensorShape.cpp:2640]: This may not be related to this commit. In line 2652, I think they mean 
```
 "entries, but got split_size=", length);
```
instead of 
```
             "entries, but got split_sizes=", split_sizes);
```

Comment (zpcore) [reply aten/src/ATen/native/TensorShape.cpp:2640]: Sorry, omit my comment. I think it is correct.

**Comment by zpcore** [comment]
Hi @andreigh , we have updated the XLA PR https://github.com/pytorch/xla/pull/5655 (hash: cf55b254c5710cfc2aba105ef326887f8030e9f1) to resolve the CI failure. Note the hash has been updated. Thanks.

File: .github/ci_commit_pins/xla.txt [hunk 5]
```diff
@@ -1 +1 @@
-b84f79803f64354a8be3af9e5b53349f65bd7494
+22b777961074110fb8101c16a2eda7ddddbf1858
```
Comment (zpcore) [inline .github/ci_commit_pins/xla.txt:1]: Can you update the hash to cf55b254c5710cfc2aba105ef326887f8030e9f1? I have made the change. This will resolve the XLA tests failing issue. Thanks.

**Comment by andreigh** [comment]
CI is green ! Who can review ? Thank you
